### PR TITLE
sql: add builtin to request statement bundles

### PR DESCRIPTION
**Pull request by Azhng**
Makes https://github.com/cockroachdb/cockroach/issues/79547 less painful.

Previously, there was no way to request statement bundle builtins for
internal statements.
This commit introduces crdb_internal.request_statement_bundle builtin
which allows statement bundle being requested from SQL CLI. The new
builtin takes three parameters:
* statement fingerprint text
* minimum execution latency for the statement
* length of duration the statement bundle request will stay valid for
VIEWACTIVITY or ADMIN role option is required to use this builtin. If user has
VIEWACTIVITYREDACTED role option, user is not allowed to use this builtin.

Release note (sql change): new crdb_internal.request_statement_bundle builtin
allows statement bundle being requested from SQL CLI. The new builtin takes
three parameters:
* statement fingerprint text
* minimum execution latency for the statement
* length of duration the statement bundle request will stay valid for.
VIEWACTIVITY or ADMIN role option is required to use this builtin. If user has
VIEWACTIVITYREDACTED role option, user is not allowed to use this builtin.

### CODE REVIEW

**Review by matthewtodd — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-:-N-Tvghl6AeNs-aqJ11D:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

File: pkg/sql/sem/builtins/builtins.go [hunk 1]
```diff
@@ -6876,6 +6876,55 @@ specified store on the node it's run from. One of 'mvccGC', 'merge', 'split',
...
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{"stmtFingerprint", types.String},
+				{"minExecutionLatency", types.Interval},
+				{"expiresAfter", types.Interval},
+			},
+			ReturnType: tree.FixedReturnType(types.Bool),
+			Fn: func(evalCtx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				hasViewActivityRedacted, err := evalCtx.SessionAccessor.HasRoleOption(
+					evalCtx.Ctx(), roleoption.VIEWACTIVITYREDACTED)
```
Comment (Azhng) [inline pkg/sql/sem/builtins/builtins.go:6894]: Note to self: this is problematic, view activity redacted should not be able to request statement bundles

**Review by maryliag — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79693)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng and @maryliag)

___
*[-- commits, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N079DKw4N-6LUKTuBDE):*
VIEWACTIVITYREDACTED should not have access to statement bundle, only VIEWACTIVITY
___
*[pkg/sql/sem/builtins/builtins.go, line 6894 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N-9rJ37BOOlhag0dBsa-r1-6894:-N079O6Q9nEzB8DQFMju:b-wq4kbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b416857fa773dc823da2f3f8ec2a63ec8700b7/pkg/sql/sem/builtins/builtins.go#L6894)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Note to self: this is problematic, view activity redacted should not be able to request statement bundles 
</blockquote></details>

from your comment, this should not be here, VIEWACTIVITYREDACTED should not have access to statement bundle, so check for admin and VIEWACTIVITY
we do have function about require VIEWACTIVITY and NOT  VIEWACTIVITYREDACTED
___
*[pkg/server/status_test.go, line 2753 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N079eMD38Nxa8AikwdG:-N079eMD38Nxa8AikwdH:bbx0tc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b416857fa773dc823da2f3f8ec2a63ec8700b7/pkg/server/status_test.go#L2753)):*
> ```Go
> 	})
> 
> 	t.Run("builtin", func(t *testing.T) {
> ```

after you make the changes to allow only admin and VIEWACTIVITY, create a test that checks those have access but that it will fail for VIEWACTIVITYREDACTED

File: pkg/sql/sem/builtins/builtins.go [hunk 2]
```diff
@@ -6876,6 +6876,55 @@ specified store on the node it's run from. One of 'mvccGC', 'merge', 'split',
...
+					stmtFingerprint,
+					minExecutionLatency,
+					expiresAfter,
+				); err != nil {
+					return nil, err
+				}
+
+				return tree.DBoolTrue, nil
+			},
+			Volatility: tree.VolatilityVolatile,
```
Comment (otan) [inline pkg/sql/sem/builtins/builtins.go:6925]: we should probably add an info field

**Review by otan — APPROVED** [review]

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79693)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag and @otan)

___
*[`-- commits` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N079DKw4N-6LUKTuBDE):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

VIEWACTIVITYREDACTED should not have access to statement bundle, only VIEWACTIVITY
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 6894 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N-9rJ37BOOlhag0dBsa-r1-6894:-N1PZY9I9oMc1KWJ_0Fm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b416857fa773dc823da2f3f8ec2a63ec8700b7/pkg/sql/sem/builtins/builtins.go#L6894)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

from your comment, this should not be here, VIEWACTIVITYREDACTED should not have access to statement bundle, so check for admin and VIEWACTIVITY
we do have function about require VIEWACTIVITY and NOT  VIEWACTIVITYREDACTED
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 6925 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N-9rJ37BOOlhag0dBsa-r1-6925:-N1PZ_v5A3J_yClDlBlE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b416857fa773dc823da2f3f8ec2a63ec8700b7/pkg/sql/sem/builtins/builtins.go#L6925)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

we should probably add an info field
</blockquote></details>

Done.
___
*[`pkg/server/status_test.go` line 2753 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N079eMD38Nxa8AikwdG:-N1PZdEJ8HlqOnUhKESq:b-7f77ca) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b416857fa773dc823da2f3f8ec2a63ec8700b7/pkg/server/status_test.go#L2753)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

after you make the changes to allow only admin and VIEWACTIVITY, create a test that checks those have access but that it will fail for VIEWACTIVITYREDACTED
</blockquote></details>

Added the check in the  `TestCreateStatementDiagnosticsReportWithViewActivityOptions` below

**Review by maryliag — COMMENTED** [review]
small nits from me

> Reviewed 2 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-:-N1PpZbH52YLQYPP2dLa:b-pkd50a)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @maryliag, and @otan)

___
*[`pkg/sql/sem/builtins/builtins.go` line 6894 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N-9rJ37BOOlhag0dBsa-r1-6894:-N1Pnv7Q0niuyM52ep9D:b-1io0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b416857fa773dc823da2f3f8ec2a63ec8700b7/pkg/sql/sem/builtins/builtins.go#L6894)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

you can move this part to just below the `if !hasViewActivity {` check, so you don't need to make this check if the user is not admin or doesn't have view activity role
___
*[`pkg/sql/sem/builtins/builtins.go` line 7004 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PpENMAi6FhpvhyP3A:-N1PpENMAi6FhpvhyP3B:b-xczlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/builtins/builtins.go#L7004)):*
> ```Go
> 				if !hasViewActivity {
> 					return nil, errors.New("requesting statement bundle requires " +
> 						"VIEWACTIVITY option")
> ```

nit: VIEWACTIVITY or ADMIN
___
*[`pkg/sql/sem/builtins/builtins.go` line 7006 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PoM9L4J87xLX-FctD:-N1PoM9L4J87xLX-FctE:b2xot63) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/builtins/builtins.go#L7006)):*
> ```Go
> 						"VIEWACTIVITY option")
> 				}
> ```

move to here
___
*[`pkg/sql/sem/builtins/builtins.go` line 7029 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PoS2vBvSOO4Y1W_20:-N1PoS2vBvSOO4Y1W_21:b-vslfds) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/builtins/builtins.go#L7029)):*
> ```Go
> 			Volatility: volatility.Volatile,
> 			Info: `Used to request statement bundle for a given statement fingerprint
> that has execution latency than the 'minExecutionLatency'. If the 'expiresAfter'
> ```

nit: ...latency greater than the...
___
*[`pkg/sql/sem/eval/context.go` line 208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PovqN11hXhUW-j-ke:-N1PovqN11hXhUW-j-kf:batiasj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/eval/context.go#L208)):*
> ```Go
> 
> 	// StmtDiagnosticsRequestInserter is used by the
> 	// crdb_internal.request_statement_bundle builtin to insert statement bundle
> ```

nit: insert a statement
___
*[`pkg/sql/sem/eval/deps.go` line 494 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1Pp-5ZD0HAUfODqRAE:-N1Pp-5ZD0HAUfODqRAF:batiasj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/eval/deps.go#L494)):*
> ```Go
> 
> // StmtDiagnosticsRequestInsertFunc is an interface embedded in EvalCtx that can
> // be used by the builtins to insert statement diagnostics request. This
> ```

nit: insert a statement
___
*[`pkg/server/status_test.go` line 2810 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PpLDX0RxM4PiDgGI_:-N1PpLDX0RxM4PiDgGIa:bts18lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/server/status_test.go#L2810)):*
> ```Go
> 		"SELECT crdb_internal.request_statement_bundle('SELECT _', 0::INTERVAL, 0::INTERVAL)",
> 	)
> 	require.Contains(t, err.Error(), "requesting statement bundle requires VIEWACTIVITY option")
> ```

nit: or ADMIN

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79693)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag and @otan)

___
*[`pkg/sql/sem/builtins/builtins.go` line 7004 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PpENMAi6FhpvhyP3A:-N1Pvmj47ban7qSYGTre:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/builtins/builtins.go#L7004)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: VIEWACTIVITY or ADMIN
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7006 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PoM9L4J87xLX-FctD:-N1PvoR6EOIJyCeMpifZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/builtins/builtins.go#L7006)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

move to here
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7029 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PoS2vBvSOO4Y1W_20:-N1PvuOi06cPE_bKmc8e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/builtins/builtins.go#L7029)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: ...latency greater than the...
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/context.go` line 208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PovqN11hXhUW-j-ke:-N1Pvyfe7khZ-rR3I6ah:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/eval/context.go#L208)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: insert a statement
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/deps.go` line 494 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1Pp-5ZD0HAUfODqRAE:-N1Pw2lA0pSygKv8Jnxq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/sql/sem/eval/deps.go#L494)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: insert a statement
</blockquote></details>

Done.
___
*[`pkg/server/status_test.go` line 2810 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-N1PpLDX0RxM4PiDgGI_:-N1PwCUB-Vf2N1PgFE9y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c51521515a88907a951582904d0acd4d3811fa/pkg/server/status_test.go#L2810)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: or ADMIN
</blockquote></details>

Done.

**Review by maryliag — APPROVED** [review]
:lgtm:

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79693#-:-N1PygAP4oUE6HI9RbLF:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag and @otan)

**Comment by Azhng** [comment]
TFTR!

bors r+

**Pull request merged**